Login: Show error screens upon app start if there was error during site login address flow previously

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #7419
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
Previously, when a user is logged in without a selected site, we're showing the store picker. In many cases, they end up with an empty store picker with an action button to install Jetpack. This screen can be misleading and is not helpful.

This PR improves this experience by adding a few checks:
- If the user encountered an issue with the site address login flow earlier, we keep the site URL in the local storage. So the next time they start the app, we check for the problem of the app and display an error screen to help them pick up where they left off (e.g account mismatched or missing Woo).
- If the user doesn't have any connected store, we log them out since the empty store picker isn't helpful. We can consider showing some other content for store creation here for further enhancement later.

Technical changes:
- Update `LoggedOutAppSettings` with a new variable for `errorLoginSiteAddress`.
- Inject `LoggedOutAppSettings` to `AuthenticationManager` from `AppCoordinator`. Since `AuthenticationManager` is a singleton, and I don't want to move `LoggedOutAppSettings` to `ServiceLocator` as well, so I use method injection here instead of initializer injection.
- Extract a method for getting the error view controller for a given site URL in `AuthenticationManager`. This method is internal so that it can be accessed by `AppCoordinator`.
- Update `AppCoordinator` to present error view controller or log out if needed when handling the logged-in state without default store.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
Testing Account Mismatch error:
1. Log out of the app if needed, and select Enter site address.
2. Enter the address for a Woo store to which you don't have ac
cess.
3. Proceed to log in with your WP.com account.
4. After the login succeeds, the Account Mismatch screen is presented. Kill the app.
5. Open the app again, notice that you get the Account Mismatch screen again. Tap Log out, you should get back to the login screen.

Testing No Woo error:
1. Log out of the app if needed, and select Enter site address.
2. Enter the address of a self-hosted site with Jetpack and no Woo.
3. Proceed to log in with your WP.com account.
4. After the login succeeds, the No Woo screen is presented. Kill the app.
5. Open the app again, notice that you get the No Woo screen again. The Install WooCommerce and Log out buttons should still work as expected.

The last case where user is logged in without neither of the above errors and has no connected store is probably an edge case. Please feel free to test it by the following steps:
- Invite a test user to your store.
- Follow the site address login flow to log in with that user.
- On the store picker screen, kill the app.
- Remove the test user from your store.
- Open the app again, the user should stay in the prologue screen now.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->
Account Mismatch error:
<img src="https://user-images.githubusercontent.com/5533851/183052618-25c6ddab-4e66-4779-94bc-d89f7ee71ae6.png" width=320 />

No Woo error:
<img src="https://user-images.githubusercontent.com/5533851/183052700-f2063d99-7164-4428-90e1-cb0c509a1965.png" width=320 />



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Add new user defaults key for error site address during login

itsmeichigo committed:
Update LoggedOutAppSettings with errorLoginSiteAddress

itsmeichigo committed:
Extract method for handling epilogue screen

itsmeichigo committed:
Make setErrorLoginSiteAddress accept optional string

itsmeichigo committed:
Update AuthenticationManager to save the error site address

itsmeichigo committed:
Handle logged in state without default store to check for error cases

itsmeichigo committed:
Make it possible to inject storage manager to ULAccountMatcher for testability

itsmeichigo committed:
Update MockLoggedOutAppSettings

itsmeichigo committed:
Inject storageManager to AuthenticationManager

itsmeichigo committed:
Update unit tests for AuthenticationManager

itsmeichigo committed:
Update unit tests for AppCoordinator

itsmeichigo committed:
Merge branch 'trunk' into issue/7419-handle-logged-out-state-for-no-stores

itsmeichigo committed:
Fix build failure

itsmeichigo committed:
Silence hound

itsmeichigo committed:
Update navigation item for error controller

jaclync commented on WooCommerce/Classes/Extensions/UserDefaults+Woo.swift:
```diff
@@ -21,6 +21,7 @@ extension UserDefaults {
         case analyticsUsername
         case notificationsLastSeenTime
         case notificationsMarkAsReadCount
+        case errorLoginSiteAddress
```
super nit: maybe we can try keeping this in alphabetical order for easier lookup, even though some of the latter ones aren't really following the order?

itsmeichigo replied:
Updated in 8d69b12cc66ad83391031a533e87a9515715b9be 👍

jaclync commented on WooCommerce/Classes/ViewRelated/AppCoordinator.swift:
```diff
@@ -166,18 +172,43 @@ private extension AppCoordinator {
+            }
+        }
+
+        // If no store is found and no error is detected, log the user out.
```
💬 for this scenario, I feel like it's okay to leave the user logged in? was there a discussion on this handling that I can read about more context? this gives them a chance to set up a store on the website for now and then come back. They can always tap "Try another account" to log out if they ever come back.

itsmeichigo replied:
There was no discussion around this, but the current empty store picker is showing the Jetpack error which is misleading so I wanted to log the user out. But maybe we should improve the No store screen instead, so I reverted the change in 20f2beed03349562dc45d20097c802fad5ef55ac.

jaclync commented on WooCommerce/Classes/ViewRelated/AppCoordinator.swift:
```diff
@@ -166,18 +172,43 @@ private extension AppCoordinator {
+        // Show error for the current site URL if exists.
+        if let siteURL = loggedOutAppSettings.errorLoginSiteAddress {
+            let authenticationUI = authenticationManager.authenticationUI()
+            if let authenticationUI = authenticationUI as? UINavigationController,
```
nit: could this be
```suggestion
            if let authenticationUI = authenticationManager.authenticationUI() as? UINavigationController,
```

itsmeichigo replied:
Updated in fa214cd4310a1de6a56b440bc07839938745bb93.

jaclync commented on WooCommerce/Classes/Authentication/AuthenticationManager.swift:
```diff
@@ -155,6 +168,43 @@ class AuthenticationManager: Authentication {
+                showsInstallButton: matchedSite.isJetpackConnected,
+                onSetupCompletion: { [weak self] siteID in
+                    guard let self = self else { return }
+                    self.startStorePicker(with: siteID, in: navigationController)
```
the previous call has a `onDismiss` parameter, I'm guessing this parameter isn't needed in this case?

itsmeichigo replied:
Good catch! We don't need the closure at app start, but it may still be necessary during the login flow, so I added it back in 82d97cb5d660db38bf6002aeee769d8bcd38def0.

jaclync commented on WooCommerce/WooCommerceTests/AppCoordinatorTests.swift:
```diff
@@ -305,6 +388,7 @@ private extension AppCoordinatorTests {
     /// Convenience method to make AppCoordinator instances.
     func makeCoordinator(window: UIWindow? = nil,
                          stores: StoresManager? = nil,
+                         storageManager: StorageManagerType? = nil,
```
nit: since `storageManager` is already initialized for all test cases and the same value is injected to `AppCoordinator` below, maybe this parameter isn't needed?

itsmeichigo replied:
Updated in 392eb668c35b4bad54e8e8dc1e2ab3b0b9a4c868.

jaclync commented on WooCommerce/WooCommerceTests/Authentication/AuthenticationManagerTests.swift:
```diff
@@ -190,4 +190,107 @@ final class AuthenticationManagerTests: XCTestCase {
         XCTAssertTrue(rootController is ULErrorViewController)
     }
+
+    func test_errorViewController_returns_account_mismatch_if_no_site_matches_the_given_url() {
```
thanks for adding these tests!

jaclync commented on WooCommerce/Classes/ViewRelated/AppCoordinator.swift:
```diff
@@ -166,18 +172,43 @@ private extension AppCoordinator {
+    /// check for errors and display store picker if none exists.
     ///
-    func displayStorePicker() {
+    func handleLoggedInStateWithoutDefaultStore() {
```
nit: maybe we can keep the `display*` to indicate that this is to display some UI?

```suggestion
    func displayLoggedInStateWithoutDefaultStore() {
```

itsmeichigo replied:
Updated in 7846969070aea77aa19b3fda0c0c05e3441c28f3.

jaclync reviewed (commented):
Changes LGTM, mostly a question about automatically logging out when the user has no stores 💭 it's exciting being able to install Woo!

> Testing Account Mismatch error:
* In this case, it might be nice if we can let the user log in to another store even if it doesn’t match the site address they entered before

<img src="https://user-images.githubusercontent.com/1945542/183333130-57a2e190-6be2-4bfd-8af6-6e52d4fe1ef9.png" width="300" />


> Testing No Woo error:
* This is most likely an edge case: I used a JN site with Jetpack only without any WP.com user at first. Because the site already has Jetpack installed and activated but no [WP.com](http://wp.com/) accounts connected to the site, it still shows the Install Jetpack screen and continuing with the Install Jetpack results in some weird state. Eventually, I was able to reach the Install Woo screen after restarting login with site address
* After installing Woo from this flow, it didn’t proceed to the store automatically. eventually, I logged in to the store after navigating back to the previous [WP.com](http://wp.com/) screen (2FA in my case)
* On the Install Woo web view, maybe we can hide the mention of “pricing” at the top? 😅 It seems irrelevant during the installation
<img src="https://user-images.githubusercontent.com/1945542/183333235-9ce29b54-2c38-49eb-afe8-a6b6c6f3b7a1.png" width="300" />

itsmeichigo committed:
Simplify code following review

itsmeichigo committed:
Keep user logged in if they have no connected stores

itsmeichigo committed:
Move `case errorLoginSiteAddress` to follow alphabetic order

itsmeichigo committed:
Rename `handleLoggedInStateWithoutDefaultStore` method

itsmeichigo committed:
Send onStorePickerDismiss to `errorViewController`

itsmeichigo committed:
Remove `storageManager` from `makeCoordinator`

itsmeichigo committed:
Unhide action buttons for store picker with configuration `.listStores`

itsmeichigo commented:
Thank you @jaclync for the thorough review! 🙇‍♀️ 

> Testing Account Mismatch error: In this case, it might be nice if we can let the user log in to another store even if it doesn’t match the site address they entered before

Great catch! Updated in 5e00936d641f98b081c9c37c73ac0a214ee75376 to always show the action buttons on the store picker with configuration `.listStores`.
(**Updated**: Also added d8598561861e6302b6d64de528d1bfc86ebf5107 to fix the issue selecting a store for this configuration).

> This is most likely an edge case: I used a JN site with Jetpack only without any [WP.com](http://wp.com/) user at first. Because the site already has Jetpack installed and activated but no [WP.com](http://wp.com/) accounts connected to the site, it still shows the Install Jetpack screen and continuing with the Install Jetpack results in some weird state. Eventually, I was able to reach the Install Woo screen after restarting login with site address

For this edge case, did you connect Jetpack manually or did you still manage to do it on the app's web view? The web view solution isn't perfect with a lot of issues, we might want to migrate to the native flow soon 😓.

> After installing Woo from this flow, it didn’t proceed to the store automatically. eventually, I logged in to the store after navigating back to the previous WP.com screen (2FA in my case)

Could you recall the last screen on the web view of the Woo install flow in this case? There can be cases where the setup flow doesn't start from scratch (e.g when Woo is installed but not activated), and our current solution of catching the completion state doesn't work. We'll need to figure out a working solution for these edge cases, maybe in a future update.

> On the Install Woo web view, maybe we can hide the mention of “pricing” at the top? 😅 It seems irrelevant during the installation

This web page is reused for all cases: non-Atomic, Atomic, and self-hosted sites - which is why the pricing header is there. We'll need to work with the frontend team to find a workaround, but for now, I think this shouldn't be a blocker.

I've addressed other issues in the comment. Please let me know if I miss anything - thanks!

itsmeichigo committed:
Fix issue handling the select store button in the store picker with configuration `.listStores`

jaclync reviewed (approved):
Thanks for the updates, LGTM!

> Great catch! Updated in https://github.com/woocommerce/woocommerce-ios/commit/5e00936d641f98b081c9c37c73ac0a214ee75376 to always show the action buttons on the store picker with configuration .listStores.
(Updated: Also added https://github.com/woocommerce/woocommerce-ios/commit/d8598561861e6302b6d64de528d1bfc86ebf5107 to fix the issue selecting a store for this configuration).

I tested it again, but I still couldn't select a different site other than the default first site 🤔 I can still tap Continue to log in to a store and then switch stores though.

> For this edge case, did you connect Jetpack manually or did you still manage to do it on the app's web view? The web view solution isn't perfect with a lot of issues, we might want to migrate to the native flow soon 😓.

Jetpack-the-plugin was already installed on the JN site (Jetpack but no Woo), so I was confused when I saw the Install Jetpack screen. I think this is shown because there is no user associated with a [WP.com](http://wp.com/) account. I just reproduced this case again, and it looks like I was able to connect Jetpack to the previously entered [WP.com](http://wp.com/) account by tapping “Install Jetpack” but I couldn’t get past this screen unless I restart the login flow.

> Could you recall the last screen on the web view of the Woo install flow in this case? There can be cases where the setup flow doesn't start from scratch (e.g when Woo is installed but not activated), and our current solution of catching the completion state doesn't work. We'll need to figure out a working solution for these edge cases, maybe in a future update.

I was able to reproduce this again, and I saw the `Cannot verify your site's WooCommerce installation.` notice this time. The installation & activation were successful though.

<img src="https://user-images.githubusercontent.com/1945542/183362426-d9b40266-c45e-4a16-8724-d051a887fd6b.png" width="300" />

> This web page is reused for all cases: non-Atomic, Atomic, and self-hosted sites - which is why the pricing header is there. We'll need to work with the frontend team to find a workaround, but for now, I think this shouldn't be a blocker.

Yea no problem, I didn't see any prices mentioned on this page so that shouldn't be a problem for App Store review.

itsmeichigo committed:
Fix issue highlighting selected store in store picker for `.listStore` configuration

itsmeichigo commented:
> I tested it again, but I still couldn't select a different site other than the default first site 🤔 I can still tap Continue to log in to a store and then switch stores though.

I missed this - fixed in 2423430c897538e43ba4eeb92d53b15f39a4f56b - thanks!

> Jetpack-the-plugin was already installed on the JN site (Jetpack but no Woo), so I was confused when I saw the Install Jetpack screen. I think this is shown because there is no user associated with a [WP.com](http://wp.com/) account. I just reproduced this case again, and it looks like I was able to connect Jetpack to the previously entered [WP.com](http://wp.com/) account by tapping “Install Jetpack” but I couldn’t get past this screen unless I restart the login flow.

Did you see the paywall when tapping Install Jetpack? I assume the web view didn't close by itself?

> I was able to reproduce this again, and I saw the Cannot verify your site's WooCommerce installation. notice this time. The installation & activation were successful though.

This could be because Jetpack sync hasn't been completed after Woo installation - I added another retry in 6b4c24886a548e9424bb21ea6e1f15170d1445fd to match [Android's behavior](https://github.com/woocommerce/woocommerce-android/blob/60137e9f818a1ced4135ca2faedaf1ecca8c99d1/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerViewModel.kt#L401) and reduce the number of failures.

itsmeichigo committed:
Add more retry for refreshing sites after Woo Installation

jaclync commented:
> Did you see the paywall when tapping Install Jetpack? I assume the web view didn't close by itself?

There are prices for each paid Jetpack plan after installing & activating the plugin, and yes the web view did not dismiss itself. This can be reproduced by:

- Set up a site with Jetpack-the-plugin installed & activated but no Woo, I used JN with the default options. No users are connected to a WP.com account
- Follow the **Testing No Woo error** testing steps in the PR description --> the Jetpack installation screen is shown instead of the Install Woo screen after entering the WP.com credentials

> I missed this - fixed in https://github.com/woocommerce/woocommerce-ios/commit/2423430c897538e43ba4eeb92d53b15f39a4f56b - thanks!

Site selection works fine now 👍

itsmeichigo commented:
Thank you @jaclync for clarifying the testing steps! This is the case that we cannot cover with the web solution since the web flow assumes a site with Jetpack installed and activated is completed and shows the paywall immediately without guiding the user to the connection step.

I'll follow up with this issue in another update for the Jetpack Install experiment - I'm merging this PR for now to make it to the code freeze. Thank you for the review!
